src/engines/julia/update_rules/nonlinear_conjugate.jl

```diff
@@ -0,0 +1,301 @@
+# Inplementation based on Khan et al. (2017), "Conjugate-computation variational inference: 
+# Converting variational inference in non-conjugate models to inferences in conjugate models",
+# and Akbayrak et al. (2021), "Extended Variational Message Passing for Automated Approximate
+# Bayesian Inference"
+
+export
+ruleSPNonlinearCOutNM,
+ruleSPNonlinearCIn1MN,
+ruleSPNonlinearCOutNMX,
+ruleSPNonlinearCInMX,
+ruleSPNonlinearCInGX,
+ruleMNonlinearCInMGX
+
+const default_optimizer = ForgetDelayDescent(200.0, 0.6) # Default optimizer
+const default_n_iterations = 1000 # Default number of iterations for gradient descent
+
+
+#-----------------------
+# Conjugate Update Rules
+#-----------------------
+
+ruleSPNonlinearCOutNM(g::Function,
+                      msg_out::Nothing,
+                      msg_in1::Message;
+                      dims::Any=nothing,
+                      n_samples=default_n_samples,
+                      n_iterations=default_n_iterations,
+                      optimizer=default_optimizer) =
+    ruleSPNonlinearSOutNM(g, nothing, msg_in1; dims=dims, n_samples=n_samples) # Reuse sampling update
+
+function ruleSPNonlinearCIn1MN(g::Function,
+                               msg_out::Message,
+                               msg_in1::Message{F, V};
+                               dims::Any=nothing,
+                               n_samples=default_n_samples,
+                               n_iterations=default_n_iterations,
+                               optimizer=default_optimizer) where {F<:FactorNode, V<:VariateType}
+
+    msg_s = ruleSPNonlinearSIn1MN(g, msg_out, nothing; dims=dims, n_samples=n_samples) # Returns Message{Function}
+    η = naturalParams(msg_in1.dist)
+    λ = renderCVI(msg_s.dist.params[:log_pdf], n_iterations, optimizer, η, msg_in1)
+
+    return Message(standardDistribution(V, F, η=λ-η))
+end
+
+ruleSPNonlinearCOutNMX(g::Function,
+                       msg_out::Nothing,
+                       msgs_in::Vararg{Message};
+                       dims::Any=nothing,
+                       n_samples=default_n_samples,
+                       n_iterations=default_n_iterations,
+                       optimizer=default_optimizer) =
+     ruleSPNonlinearSOutNMX(g, nothing, msgs_in...; dims=dims, n_samples=n_samples)                                 
+
+function ruleSPNonlinearCInGX(g::Function,
```

Comment:
This implementation assumes no factorization in input variables. This must be stated at least as a comment. Besides, it restricts the applicability of the method and does not allow multi mixed distributions. It is weird because in forward direction ruleSPNonlinearCOutNMX supports message calculations with mixed distribution types. But there is no rule to support backward message calculations. I suggest we assume mean-field factorization as in original implementation to support borader range of cases, or revise ruleSPNonlinearCOutNMX in such a way that only Gaussian inputs are allowed.